app/uk/gov/hmrc/customsexportsinternalfrontend/controllers/actions/AuthenticatedAction.scala

```diff
@@ -0,0 +1,72 @@
+/*
+ * Copyright 2019 HM Revenue & Customs
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package uk.gov.hmrc.customsexportsinternalfrontend.controllers.actions
+
+import javax.inject.Inject
+import javax.inject.Singleton
+import play.api.mvc.Results.Forbidden
+import play.api.mvc._
+import play.api.{Configuration, Environment, Logger}
+import uk.gov.hmrc.auth.core.AuthProvider.PrivilegedApplication
+import uk.gov.hmrc.auth.core.retrieve.v2.Retrievals
+import uk.gov.hmrc.auth.core.{AuthProviders, AuthorisationException, AuthorisedFunctions, NoActiveSession}
+import uk.gov.hmrc.customsexportsinternalfrontend.config.AppConfig
+import uk.gov.hmrc.customsexportsinternalfrontend.connectors.StrideAuthConnector
+import uk.gov.hmrc.customsexportsinternalfrontend.controllers.exchanges.{AuthenticatedRequest, Operator}
+import uk.gov.hmrc.customsexportsinternalfrontend.views.html.unauthorized
+import uk.gov.hmrc.http.HeaderCarrier
+import uk.gov.hmrc.play.HeaderCarrierConverter
+import uk.gov.hmrc.play.bootstrap.config.AuthRedirects
+
+import scala.concurrent.{ExecutionContext, Future}
+
+@Singleton
+class AuthenticatedAction @Inject()(
+  appConfig: AppConfig,
+  override val config: Configuration,
+  override val env: Environment,
+  override val authConnector: StrideAuthConnector,
+  mcc: MessagesControllerComponents,
+  unauthorizedPage: unauthorized
+) extends ActionBuilder[AuthenticatedRequest, AnyContent] with AuthorisedFunctions with AuthRedirects {
```

Comment:
IMO unauthorisedPage should be handled not in the action but in the controller.

MessagesControllerComponents is for Controller not for Action. It's necessary for FrontendController.

IMO This AutheticatedAction is wrong, is not doing what expected. Doing more things than should.

Reply:
What would you suggest?

Reply:
Logic and view in Controller/Handler etc, The same approach like we have in other services. Above examples for movements and exports. Movements are closer to correct implementation if you look for the approaches in new play version. (play 2.6 and newer)


Example of unauthorised usage:

            Future.successful(Results.Redirect(routes.UnauthorisedController.onPageLoad()

Controller where we will have logic around unauthorised. If you will have another place where you need to go to unauthorised page with this approach you still need to inject view and everything. 

Unauthorised is treated as normal endpoint in every our service. Implementation in this PR is not based on our services.

Reply:
The thing is the UnauthorisedController in those services is just a page with an error message, it doesn't actually mean you are unauthorised. I could be logged in and go to `/unauthorised` for whatever reason and it would incorrectly tell me I am not unauthorised. A little weird.

The way I've implemented it more closely mirrors how the `error` view is used:
- If you access an endpoint that errors you get the error view (not a generic redirect)
Similarly with the way this is implemented:
- If you access an endpoint that you are not allowed to you get an unauthorised view (not a generic redirect)

I don't think there is anything wrong with Actions loading views, it saves us having unnecessary endpoints with generic error messages

Reply:
If you go to unauthorised you're expecting that you'll have something related with unauthorised.

In my opinion Action and Controller are two different things. Everyone has different approaches. You can't change my thinking about Action, Controllers. I said everything that I can around those two things. Knowledge and approach based on the documentation and another our services. Still will repeat it, I'm not blocking this, but I won't approve it too.

This discussion doesn't make sense in my opinion, it's going to the wrong direction. It's not necessary :)

If team is happy stay with this, I'm not blocking it :)